Delphi 401 delphi task reporting stats optimisation patric dosch

patric-dosch-vi opened the pull request:
This is the version using mongo aggregations to get the count data faster.

patric-dosch-vi committed:
feat: Rework registration statistics to use mongo aggregations for the additional data, i.e. counts for messages, calls, etc.

The idea of implementing the registration statistic in a query with multiple aggregations was rejected after initial testing in the NoSQL client.

Refs: DELPHI-401

patric-dosch-vi committed:
fix: bump actions/cache to v4

Refs: DELPHI-401

patric-dosch-vi committed:
fix: correct usage of StatisticContainer

Refs: DELPHI-401

vi-sudo commented:
Task linked: [OB-21462 [Delphi] Task: Reporting/Stats optimisation](https://app.clickup.com/t/6656196/OB-21462)

patric-dosch-vi committed:
feat: add last interaction date as the max of messageDate, videoCallDate or bookingDate

The existing aggregations are extended by the max date i.e. the date of the newest event. Instead of just a count, a POJO is now returned that contains the count and lastInteraction. In the converter layer, the latest of the three max dates is determined and used.

Refs: DELPHI-401

patric-dosch-vi commented on src/main/java/de/caritas/cob/statisticsservice/api/statistics/repository/StatisticsEventRepository.java:
```diff
@@ -77,45 +86,113 @@ class Duration {
+      pipeline = {
+        "{ '$match': { 'eventType': 'ARCHIVE_SESSION' } }",
+        "{ '$sort': { 'timestamp': -1 } }",
+        "{ '$group': {'_id': '$sessionId', 'endDate': { '$first': '$metaData.endDate' } } }"
```
lieber $max statt sort und first?

patric-dosch-vi replied:
würde ich jetzt so lassen

patric-dosch-vi commented on src/main/java/de/caritas/cob/statisticsservice/api/statistics/repository/projection/UserEventAggregation.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Getter;
+
+@Data
+@Getter
```
getter nicht schon in Data?

patric-dosch-vi replied:
@tobias-karrer @Value ist nice, ist ein immutable jetzt ;)

patric-dosch-vi committed:
fix: use @Value for immutable objects (projection)

Refs: DELPHI-401

patric-dosch-vi commented on src/main/java/de/caritas/cob/statisticsservice/api/statistics/repository/projection/SessionArchiveDate.java:
```diff
@@ -0,0 +1,9 @@
+
+import lombok.Value;
+
+@Value
```
@tobias-karrer @Value ist nice, ist ein immutable jetzt ;)
